plugins/plugin_utils/inventory.py

```diff
@@ -0,0 +1,162 @@
+            if templar.is_template(self.get_option(credential_type)):
+                credentials[credential_type] = templar.template(variable=self.get_option(credential_type), disable_lookups=False)
+            else:
+                credentials[credential_type] = self.get_option(credential_type)
+
+        self.boto_profile = credentials['aws_profile']
+        self.aws_access_key_id = credentials['aws_access_key']
+        self.aws_secret_access_key = credentials['aws_secret_key']
+        self.aws_security_token = credentials['aws_security_token']
+        self.iam_role_arn = credentials['iam_role_arn']
+
+        if not self.boto_profile and not (self.aws_access_key_id and self.aws_secret_access_key):
+            session = botocore.session.get_session()
+            try:
+                credentials = session.get_credentials().get_frozen_credentials()
+            except AttributeError:
+                pass
+            else:
+                self.aws_access_key_id = credentials.access_key
+                self.aws_secret_access_key = credentials.secret_key
+                self.aws_security_token = credentials.token
+
+        if not self.boto_profile and not (self.aws_access_key_id and self.aws_secret_access_key):
+            self.fail_aws("Insufficient boto credentials found. Please provide them in your "
+                          "inventory configuration file or set them as environment variables.")
+
+    def _boto3_assume_role(self, credentials, resource, region=None):
+        """
+        Assume an IAM role passed by iam_role_arn parameter
+
+        :return: a dict containing the credentials of the assumed role
+        """
+
+        iam_role_arn = self.iam_role_arn
+
+        try:
+            sts_connection = boto3.session.Session(profile_name=self.boto_profile).client('sts', region, **credentials)
+            role_session_name = f"ansible_aws_{resource}_dynamic_inventory"
+            sts_session = sts_connection.assume_role(RoleArn=iam_role_arn, RoleSessionName=role_session_name)
+            return dict(
+                aws_access_key_id=sts_session['Credentials']['AccessKeyId'],
+                aws_secret_access_key=sts_session['Credentials']['SecretAccessKey'],
+                aws_session_token=sts_session['Credentials']['SessionToken']
+            )
+        except botocore.exceptions.ClientError as e:
+            self.fail_aws("Unable to assume IAM role: %s" % to_native(e))
+
+    def _get_connection(self, credentials, resource, region='us-east-1'):
+        return boto3_conn(self, conn_type='client', resource=resource, region=region, **credentials)
+
+    def _boto3_conn(self, regions, resource):
+        '''
+            :param regions: A list of regions to create a boto3 client
+
+            Generator that yields a boto3 client and the region
+        '''
+
+        credentials = self._get_credentials()
+        iam_role_arn = self.iam_role_arn
+
+        if not regions:
+            try:
```

Comment:
If we're explicitly passed iam_assume_role, it kinda feels like we should be trying the assumed role first, and falling back as an act of last resort.